### PR TITLE
Bluekeep (CVE-2019-0708): add TLS, document blobs, refactor

### PR DESCRIPTION
**Pull request by TomSellers**
This PR makes changes to the BlueKeep ( CVE-2019-0708 ) auxiliary scanner module which:
- Adds TLS support which should enable checking servers which require TLS to connect.
- Documents _most_ of the binary blobs which should enable easier improvements such as setting user names, hostnames, and IPs.
- **UPDATED** - Implements user control of username, hostname, domain name, and IP address. This should reduce the ability to fingerprint this module and allow testers to better blend in with the target network.
- Refactors the code to remove duplication and prepare for possible migration of code to a RDP library.

### Known issues
1. **FIXED** - If Server 2008 is configured to only allow RDP Security this will fail to fall back. I know the root cause and will address tonight or tomorrow. 
1. **FIXED** - This PR doesn't **yet** change the username, hostname, or IP. I may go ahead an add it in this PR depending on the feedback and the time I have available to work on this PR.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `scanner/rdp/cve_2019_0708_bluekeep`
- [ ] Set `RHOSTS` as needed
- [ ] Set `VERBOSE` as needed
- [ ] **Verify** the correct vulnerability state is output when run
- [ ] **Verify** the thing does not do what it should not

### Windows XP SP3, UNpatched, RDP Security
```shell
[*] 192.168.50.137:3389   - Verifying RDP protocol...
[*] 192.168.50.137:3389   - Server requests RDP Security
[*] 192.168.50.137:3389   - Sending erect domain request
[*] 192.168.50.137:3389   - Sending security exchange PDU
[*] 192.168.50.137:3389   - Sending client info PDU
[*] 192.168.50.137:3389   - Received License packet
[*] 192.168.50.137:3389   - Received Server Demand packet
[*] 192.168.50.137:3389   - Sending client confirm active PDU
[*] 192.168.50.137:3389   - Sending client synchronize PDU
[*] 192.168.50.137:3389   - Sending client control cooperate PDU
[*] 192.168.50.137:3389   - Sending client control request control PDU
[*] 192.168.50.137:3389   - Sending client persistent key list PDU
[*] 192.168.50.137:3389   - Sending client font list PDU
[+] 192.168.50.137:3389   - The target is vulnerable.                 <-----------
[*] 192.168.50.137:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

### Windows XP SP3, patched, RDP Security
```shell
[*] 192.168.50.196:3389   - Verifying RDP protocol...
[*] 192.168.50.196:3389   - Server requests RDP Security
[*] 192.168.50.196:3389   - Sending erect domain request
[*] 192.168.50.196:3389   - Sending security exchange PDU
[*] 192.168.50.196:3389   - Sending client info PDU
[*] 192.168.50.196:3389   - Received License packet
[*] 192.168.50.196:3389   - Received Server Demand packet
[*] 192.168.50.196:3389   - Sending client confirm active PDU
[*] 192.168.50.196:3389   - Sending client synchronize PDU
[*] 192.168.50.196:3389   - Sending client control cooperate PDU
[*] 192.168.50.196:3389   - Sending client control request control PDU
[*] 192.168.50.196:3389   - Sending client persistent key list PDU
[*] 192.168.50.196:3389   - Sending client font list PDU
[*] 192.168.50.196:3389   - The target is not exploitable.                 <-----------
[*] 192.168.50.196:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```
### Windows Server 2008 Service Pack 2, UNpatched, Security Layer: RDP Security
 This is a *bug* that I'll fix before it lands **FIXED**
```shell
[*] 192.168.50.129:3389   - Verifying RDP protocol...
[*] 192.168.50.129:3389   - Server requests NLA security which mitigates this vulnerability.
[*] 192.168.50.129:3389   - The target is not exploitable.
[*] 192.168.50.129:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```
### Windows Server 2008 Service Pack 2, UNpatched, Security Layer: Negotiate
```shell
[*] 192.168.50.129:3389   - Verifying RDP protocol...
[*] 192.168.50.129:3389   - Server requests TLS                 <-----------
[*] 192.168.50.129:3389   - Sending erect domain request
[*] 192.168.50.129:3389   - Sending client info PDU
[*] 192.168.50.129:3389   - Received License packet
[*] 192.168.50.129:3389   - Received Server Demand packet
[*] 192.168.50.129:3389   - Sending client confirm active PDU
[*] 192.168.50.129:3389   - Sending client synchronize PDU
[*] 192.168.50.129:3389   - Sending client control cooperate PDU
[*] 192.168.50.129:3389   - Sending client control request control PDU
[*] 192.168.50.129:3389   - Sending client persistent key list PDU
[*] 192.168.50.129:3389   - Sending client font list PDU
[+] 192.168.50.129:3389   - The target is vulnerable.                 <-----------
[*] 192.168.50.129:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

### Windows Server 2008 Service Pack 2, UNpatched, Security Layer: SSL (TLS 1.0)
```shell
[*] 192.168.50.129:3389   - Verifying RDP protocol...
[*] 192.168.50.129:3389   - Server requests TLS                 <-----------
[*] 192.168.50.129:3389   - Sending erect domain request
[*] 192.168.50.129:3389   - Sending client info PDU
[*] 192.168.50.129:3389   - Received License packet
[*] 192.168.50.129:3389   - Received Server Demand packet
[*] 192.168.50.129:3389   - Sending client confirm active PDU
[*] 192.168.50.129:3389   - Sending client synchronize PDU
[*] 192.168.50.129:3389   - Sending client control cooperate PDU
[*] 192.168.50.129:3389   - Sending client control request control PDU
[*] 192.168.50.129:3389   - Sending client persistent key list PDU
[*] 192.168.50.129:3389   - Sending client font list PDU
[+] 192.168.50.129:3389   - The target is vulnerable.                 <-----------
[*] 192.168.50.129:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed

```

### Windows Server 2008 Service Pack 2, UNpatched, NLA required
```shell
[*] 192.168.50.129:3389   - Verifying RDP protocol...
[*] 192.168.50.129:3389   - Server requests NLA security which mitigates this vulnerability.                <-----------
[*] 192.168.50.129:3389   - The target is not exploitable.                <-----------
[*] 192.168.50.129:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

### Windows Server 2016, NLA OPTIONAL (TLS)
```shell
[*] 192.168.50.143:3389   - Verifying RDP protocol...
[*] 192.168.50.143:3389   - Server requests TLS                <-----------
[*] 192.168.50.143:3389   - Sending erect domain request
[*] 192.168.50.143:3389   - Sending client info PDU
[*] 192.168.50.143:3389   - Received License packet
[*] 192.168.50.143:3389   - Received Server Demand packet
[*] 192.168.50.143:3389   - Sending client confirm active PDU
[*] 192.168.50.143:3389   - Sending client synchronize PDU
[*] 192.168.50.143:3389   - Sending client control cooperate PDU
[*] 192.168.50.143:3389   - Sending client control request control PDU
[*] 192.168.50.143:3389   - Sending client persistent key list PDU
[*] 192.168.50.143:3389   - Sending client font list PDU
[*] 192.168.50.143:3389   - The target is not exploitable.                <-----------
[*] 192.168.50.143:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```
### Windows Server 2016, NLA REQUIRED
```shell
[*] 192.168.50.143:3389   - Verifying RDP protocol...
[*] 192.168.50.143:3389   - Server requests NLA security which mitigates this vulnerability.                <-----------
[*] 192.168.50.143:3389   - The target is not exploitable.                <-----------
[*] 192.168.50.143:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

### Windows Server 2019, NLA OPTIONAL (TLS)
```shell
[*] 192.168.50.105:3389   - Verifying RDP protocol...
[*] 192.168.50.105:3389   - Server requests TLS              <-----------
[*] 192.168.50.105:3389   - Sending erect domain request
[*] 192.168.50.105:3389   - Sending client info PDU
[*] 192.168.50.105:3389   - Received License packet
[*] 192.168.50.105:3389   - Received Server Demand packet
[*] 192.168.50.105:3389   - Sending client confirm active PDU
[*] 192.168.50.105:3389   - Sending client synchronize PDU
[*] 192.168.50.105:3389   - Sending client control cooperate PDU
[*] 192.168.50.105:3389   - Sending client control request control PDU
[*] 192.168.50.105:3389   - Sending client persistent key list PDU
[*] 192.168.50.105:3389   - Sending client font list PDU
[*] 192.168.50.105:3389   - The target is not exploitable.              <-----------
[*] 192.168.50.105:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```
### Windows Server 2019, NLA REQUIRED
```shell
[*] 192.168.50.105:3389   - Verifying RDP protocol...
[*] 192.168.50.105:3389   - Server requests NLA security which mitigates this vulnerability.              <-----------
[*] 192.168.50.105:3389   - The target is not exploitable.              <-----------
[*] 192.168.50.105:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

### CODE REVIEW

**Comment by TomSellers** [comment]
@wvu-r7 FYI

**Comment by tsellers-r7** [comment]
I tried to move anything that looked like a candidate for a RDP library to the end.

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb [hunk 1]
```diff
@@ -52,333 +53,196 @@ def report_goods
...
-    pkt << [udl_with_flag].pack("S>")
-    pkt << flags #{}"\x48\x00" # flags  SEC_INFO_PKT | SEC_ENCRYPT
-    pkt << flagsHi # flagsHi
-    pkt << rdp_hmac(hmackey, data)[0..7]
-    pkt << rdp_rc4_crypt(rc4enckey, data)
+    # return true if the response is a X.224 Connect Confirm
+    # We can't use a check for RDP Negotiation Response because WinXP excludes it
+    if res && res.match("\x03\x00\x00..\xd0")
+      server_selected_proto = 0
+      server_selected_proto = res[15..18].unpack("L<")[0] if res.length >= 19
```
Comment (wvu-r7) [inline modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: # 1

For my notes.

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb [hunk 2]
```diff
@@ -52,333 +53,196 @@ def report_goods
...
       vprint_status "Could not connect to RDP."
       return Exploit::CheckCode::Unknown
     end
 
-    # send initial client data
-    res = rdp_send_recv(pdu_connect_initial)
-    rsmod, rsexp, rsran, server_rand, bitlen = rdp_parse_serverdata(res)
+    if server_selected_proto == 1
+      vprint_status("Server requests TLS")
+      swap_sock_plain_to_ssl(nsock)
```
Comment (wvu-r7) [inline modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: # 2

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb [hunk 3]
```diff
@@ -388,58 +252,123 @@ def check_rdp_vuln
     result
   end
 
-  def check_host(ip)
-    # The check command will call this method instead of run_host
+  # Create a new SSL session on the existing socket.
+  # Stolen from exploit/smtp_deliver.rb
+  def swap_sock_plain_to_ssl(nsock)
```
Comment (wvu-r7) [inline modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: # 3

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb (hunk 2 above)
Comment (bcoles) [reply modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: for the show

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb (hunk 3 above)
Comment (bcoles) [reply modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: to get ready

**Comment by TomSellers** [comment]
I've addressed the bug where the code wouldn't correctly renegotiate if the target was configured to force RDP Standard Security.

### Windows Server 2008 Service Pack 2, UNpatched, Security Layer: RDP Security
```shell
[*] 192.168.50.129:3389   - Verifying RDP protocol...
[*] 192.168.50.129:3389   - Attempting to connect using TLS security
[*] 192.168.50.129:3389   - Attempt to connect with TLS failed with error: SSL_NOT_ALLOWED_BY_SERVER             <-----------
[*] 192.168.50.129:3389   - Attempting to connect using Standard RDP security             <-----------
[*] 192.168.50.129:3389   - Server requests RDP Security
[*] 192.168.50.129:3389   - Sending erect domain request
[*] 192.168.50.129:3389   - Sending security exchange PDU
[*] 192.168.50.129:3389   - Sending client info PDU
[*] 192.168.50.129:3389   - Received License packet
[*] 192.168.50.129:3389   - Received Server Demand packet
[*] 192.168.50.129:3389   - Sending client confirm active PDU
[*] 192.168.50.129:3389   - Sending client synchronize PDU
[*] 192.168.50.129:3389   - Sending client control cooperate PDU
[*] 192.168.50.129:3389   - Sending client control request control PDU
[*] 192.168.50.129:3389   - Sending client persistent key list PDU
[*] 192.168.50.129:3389   - Sending client font list PDU
[+] 192.168.50.129:3389   - The target is vulnerable.             <-----------
[*] 192.168.50.129:3389   - Scanned 1 of 1 hosts (100% complete)

```

There is a change in the output as well that indicates when dealing with Windows XP which doesn't return a full Negotiation Response packet.

### Windows XP SP3, UNpatched, RDP Security
```shell
[*] 192.168.50.137:3389   - Verifying RDP protocol...
[*] 192.168.50.137:3389   - Attempting to connect using TLS security
[*] 192.168.50.137:3389   - Attempt to connect with TLS failed but looks like the target is Windows XP            <-----------
[*] 192.168.50.137:3389   - Attempting to connect using Standard RDP security            <-----------
[*] 192.168.50.137:3389   - Server requests RDP Security
[*] 192.168.50.137:3389   - Sending erect domain request
[*] 192.168.50.137:3389   - Sending security exchange PDU
[*] 192.168.50.137:3389   - Sending client info PDU
[*] 192.168.50.137:3389   - Received License packet
[*] 192.168.50.137:3389   - Received Server Demand packet
[*] 192.168.50.137:3389   - Sending client confirm active PDU
[*] 192.168.50.137:3389   - Sending client synchronize PDU
[*] 192.168.50.137:3389   - Sending client control cooperate PDU
[*] 192.168.50.137:3389   - Sending client control request control PDU
[*] 192.168.50.137:3389   - Sending client persistent key list PDU
[*] 192.168.50.137:3389   - Sending client font list PDU
[+] 192.168.50.137:3389   - The target is vulnerable.            <-----------
[*] 192.168.50.137:3389   - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb [hunk 4]
```diff
@@ -108,22 +108,46 @@ def check_host(ip)
   end
 
   def check_rdp
+
```
Comment (wvu-r7) [inline modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: Not sure you meant to add this blank line.

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb [hunk 5]
```diff
@@ -564,6 +580,340 @@ def int_to_bytestring( daInt, num_chars=nil )
...
```
Comment (tsellers-r7) [inline modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: @zerosum0x0 - The payload is a Client Input Event (TS_INPUT_PDU - 2.2.8.1.1.3) with messageType Synchronize Event (TS_SYNC_EVENT - 2.2.8.1.1.3.1.5) which shouldn't have (keyboard/mouse) input data. I'm unpacking that payload now and I don't see any reason for the the 800 or so `a` bytes. Does this influence the stability or patch detection?

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb [hunk 6]
```diff
@@ -564,6 +580,489 @@ def int_to_bytestring( daInt, num_chars=nil )
...
+    pdu = "\x01\x00" +   # numEvents: 1
+    "\x00\x00" +         # pad2Octets
+    "\x00\x00\x00\x00" + # eventTime
+    "\x00\x00" +         # messageType: 0 - INPUT_EVENT_SYNC
+    # TS_SYNC_EVENT 202.8.1.1.3.1.1.5
+    "\x00\x00" +         # pad2Octets
+    "\x00\x00\x00\x00" + # toggleFlags
+    # The data below is not part of the specification
+    # Testing without the data did not discover any impact on reliability
+    # or target stablility. Will likely be removed in a future PR.
```
Comment (tsellers-r7) [inline modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: @zerosum0x0 - The payload is a Client Input Event (TS_INPUT_PDU - 2.2.8.1.1.3) with messageType Synchronize Event (TS_SYNC_EVENT - 2.2.8.1.1.3.1.5) which shouldn't have (keyboard/mouse) input data.  I don't see any reason (from a protocol perspective) for the the 800 or so a bytes. Does this influence the stability or patch detection?

Comment (zerosum0x0) [reply modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: I don't think it matters. It's just a paste from what rdesktop was saying to the server. Probably has fat that can be trimmed, would need to dig into the message syntax.

Comment (tsellers-r7) [reply modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: Thanks for the response.. I can't reproduce it and It doesn't seem to impact the stability of the check. I've trimmed the data out.

**Comment by zerosum0x0** [comment]
@TomSellers You might want to also add something similar to this fix I did to receive one full TPKT at a time.

https://github.com/RiskSense-Ops/metasploit-framework/commit/59e815f230b6a2260b8bd17ae8119398cedfb840

**Comment by tsellers-r7** [comment]
> @TomSellers You might want to also add something similar to this fix I did to receive one full TPKT at a time.
> 
> [RiskSense-Ops@59e815f](https://github.com/RiskSense-Ops/metasploit-framework/commit/59e815f230b6a2260b8bd17ae8119398cedfb840)

Thanks for the heads up on this. It breaks a bit on the fast path PDUs which aren't wrapped in TPKT but I can address that.

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb [hunk 7]
```diff
@@ -19,7 +19,8 @@ def initialize(info = {})
       'Author'         =>
         [
           'JaGoTu',
-          'zerosum0x0'
+          'zerosum0x0',
+          'Tom Sellers'
```
Comment (zerosum0x0) [inline modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: Should also credit The UK's National Cyber Security Centre (NCSC) for finding the bug.

**Comment by wvu-r7** [comment]
Hey, @tsellers-r7. Did you need any help working on this?

**Comment by TomSellers** [comment]
@wvu-r7  - Thanks, I've been traveling and unable to work on it. Just pushed some changes to allow users to configure username, client name, etc.  I need to retest against my lab.  If you have any feedback I'm probably ready for it.  Thanks much

**Comment by TomSellers** [comment]
I've implemented user control of username, hostname, domain name, and IP address. This should reduce the ability to fingerprint this module and allow testers to better blend in with the target network.

The client random value used during RDP Security is now a random value.

File: modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb (hunk 7 above)
Comment (wvu-r7) [reply modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb]: Took care of this.

**Comment by wvu-r7** [comment]
```
msf5 > use bluekeep

Matching Modules
================

   #  Name                                          Disclosure Date  Rank    Check  Description
   -  ----                                          ---------------  ----    -----  -----------
   0  auxiliary/scanner/rdp/cve_2019_0708_bluekeep  2019-05-14       normal  Yes    CVE-2019-0708 BlueKeep Microsoft Remote Desktop RCE Check


[*] Using auxiliary/scanner/rdp/cve_2019_0708_bluekeep
msf5 auxiliary(scanner/rdp/cve_2019_0708_bluekeep) > set rhosts 192.168.56.101,105
rhosts => 192.168.56.101,105
msf5 auxiliary(scanner/rdp/cve_2019_0708_bluekeep) > options

Module options (auxiliary/scanner/rdp/cve_2019_0708_bluekeep):

   Name             Current Setting     Required  Description
   ----             ---------------     --------  -----------
   RDP_CLIENT_IP    192.168.0.100       yes       The client IPv4 address to report during connect
   RDP_CLIENT_NAME  rdesktop            no        The client computer name to report during connect, UNSET = random
   RDP_DOMAIN                           no        The client domain name to report during connect
   RDP_USER                             no        The username to report during connect, UNSET = random
   RHOSTS           192.168.56.101,105  yes       The target address range or CIDR identifier
   RPORT            3389                yes       The target port (TCP)
   THREADS          1                   yes       The number of concurrent threads

msf5 auxiliary(scanner/rdp/cve_2019_0708_bluekeep) > run

[*] 192.168.56.101:3389   - Verifying RDP protocol...
[*] 192.168.56.101:3389   - Attempting to connect using TLS security
[*] 192.168.56.101:3389   - Server requests TLS
[*] 192.168.56.101:3389   - Sending erect domain request
[*] 192.168.56.101:3389   - Sending client info PDU
[*] 192.168.56.101:3389   - Received License packet
[*] 192.168.56.101:3389   - Waiting for Server Demand packet
[*] 192.168.56.101:3389   - Received Server Demand packet
[*] 192.168.56.101:3389   - Sending client confirm active PDU
[*] 192.168.56.101:3389   - Sending client synchronize PDU
[*] 192.168.56.101:3389   - Sending client control cooperate PDU
[*] 192.168.56.101:3389   - Sending client control request control PDU
[*] 192.168.56.101:3389   - Sending client input sychronize PDU
[*] 192.168.56.101:3389   - Sending client font list PDU
[*] 192.168.56.101:3389   - Sending patch check payloads
[*] 192.168.56.101:3389   - The target is not exploitable.
[*] 192.168.56.101,105:3389 - Scanned 1 of 2 hosts (50% complete)
[*] 192.168.56.105:3389   - Verifying RDP protocol...
[*] 192.168.56.105:3389   - Attempting to connect using TLS security
[*] 192.168.56.105:3389   - Server requests TLS
[*] 192.168.56.105:3389   - Sending erect domain request
[*] 192.168.56.105:3389   - Sending client info PDU
[*] 192.168.56.105:3389   - Received License packet
[*] 192.168.56.105:3389   - Waiting for Server Demand packet
[*] 192.168.56.105:3389   - Received Server Demand packet
[*] 192.168.56.105:3389   - Sending client confirm active PDU
[*] 192.168.56.105:3389   - Sending client synchronize PDU
[*] 192.168.56.105:3389   - Sending client control cooperate PDU
[*] 192.168.56.105:3389   - Sending client control request control PDU
[*] 192.168.56.105:3389   - Sending client input sychronize PDU
[*] 192.168.56.105:3389   - Sending client font list PDU
[*] 192.168.56.105:3389   - Sending patch check payloads
[+] 192.168.56.105:3389   - The target is vulnerable.
[*] 192.168.56.101,105:3389 - Scanned 2 of 2 hosts (100% complete)
[*] Auxiliary module execution completed
msf5 auxiliary(scanner/rdp/cve_2019_0708_bluekeep) >
```

**Comment by wvu-r7** [comment]
# Release Notes

This adds TLS support and documented packets to the CVE-2019-0708 (BlueKeep) scanner module.

**Pull request merged**